unixfs/hamt/hamt.go

```diff
@@ -284,33 +284,124 @@ func (ds *Shard) getChild(ctx context.Context, i int) (child, error) {
 // loadChild reads the i'th child node of this shard from disk and returns it
 // as a 'child' interface
 func (ds *Shard) loadChild(ctx context.Context, i int) (child, error) {
+	c, err := ds.loadChildValue(i)
+	if err != nil {
+		return nil, err
+	}
+	if c != nil {
+		return c, nil
+	}
+
+	lnk := ds.nd.Links()[i]
+	nd, err := lnk.GetNode(ctx, ds.dserv)
+	if err != nil {
+		return nil, err
+	}
+	c, err = NewHamtFromDag(ds.dserv, nd)
+	if err != nil {
+		return nil, err
+	}
+
+	ds.children[i] = c
+	return c, nil
+}
+
+func (ds *Shard) missingChildShards() []*cid.Cid {
+	if len(ds.children) != len(ds.nd.Links()) {
+		panic("inconsistent lengths between children array and Links array")
+	}
+	res := make([]*cid.Cid, 0, len(ds.children))
+	for i, c := range ds.children {
+		if c != nil {
+			continue
+		}
+		lnk := ds.nd.Links()[i]
+		if len(lnk.Name) == ds.maxpadlen {
+			res = append(res, lnk.Cid)
+		}
+	}
+	return res
+}
+
+// loadChild returns i'th child node if it is a 'shardValue' otherwise
+// returns nil.  If neither a child or an error is returned it is safe
+// to assume the child is also a hamt shard.
+func (ds *Shard) loadChildValue(i int) (child, error) {
 	lnk := ds.nd.Links()[i]
 	if len(lnk.Name) < ds.maxpadlen {
 		return nil, fmt.Errorf("invalid link name '%s'", lnk.Name)
 	}
 
 	var c child
 	if len(lnk.Name) == ds.maxpadlen {
-		nd, err := lnk.GetNode(ctx, ds.dserv)
-		if err != nil {
-			return nil, err
+		return nil, nil
+	}
+	lnk2 := *lnk
+	c = &shardValue{
+		key: lnk.Name[ds.maxpadlen:],
+		val: &lnk2,
+	}
+
+	ds.children[i] = c
+	return c, nil
+}
+
+// preloadChildren populates the 'children' array if some child shards
+// are not already loaded they are fetched in parallel using GetMany
+func (ds *Shard) preloadChildren(ctx context.Context, f *fetcher) error {
+	if len(ds.children) != len(ds.nd.Links()) {
+		return fmt.Errorf("inconsistent lengths between children array and Links array")
```

Comment:
feels inconsistent to panic over this in one place, but throw an error here.

Reply:
Fixed.